### PR TITLE
Update programming guide to suggest using NewDoFn approach to paramet…

### PR DESCRIPTION
**Pull request by reuvenlax**
Update documentation to suggest new parameter style instead of ProcessContext.

### CODE REVIEW

File: src/documentation/programming-guide.md [hunk 1]
```diff
@@ -684,8 +684,8 @@ PCollection<Integer> wordLengths = words.apply(
   "ComputeWordLengths",                     // the transform name
   ParDo.of(new DoFn<String, Integer>() {    // a DoFn as an anonymous inner class instance
       @ProcessElement
-      public void processElement(ProcessContext c) {
-        c.output(c.element().length());
+      public void processElement(@Element String word, OutputReceiver out) {
```
Comment (jkff) [inline src/documentation/programming-guide.md]: `OutputReceiver<String>`

File: src/documentation/programming-guide.md [hunk 2]
```diff
@@ -1251,6 +1251,8 @@ In general, your user code must fulfill at least these requirements:
   Beam SDKs are not thread-safe*.
 
 In addition, it's recommended that you make your function object **idempotent**.
+Non-idempotent functions are supported by Beam, but xN require additional
```
Comment (jkff) [inline src/documentation/programming-guide.md]: xN?

File: src/documentation/programming-guide.md [hunk 3]
```diff
@@ -1504,6 +1504,61 @@ together.
...
+
+```java
+.of(new DoFn<String, String>() {
+     public void processElement(@Element String word, PaneInfo paneInfo) {
+  }})
+```
+
+{:.language-java}
+**PipelineOptions:**
+The `PipelineOptions` for the current pipeline can always be accessed in a process method by adding it as a paramet:
```
Comment (jkff) [inline src/documentation/programming-guide.md]: parameter

**Comment by reuvenlax** [comment]
@melap How can I generate the docs to see what this looks like?

File: src/documentation/programming-guide.md (hunk 1 above)
Comment (iemejia) [reply src/documentation/programming-guide.md]: Is this backwards compatible to previous releases of Beam (2.x)  if not, maybe we should say explictly so.

Comment (reuvenlax) [reply src/documentation/programming-guide.md]: @iemejia yes it is. ProcessContext still exists as before, and cannot be removed until Beam 3.0 for compatibility reasons.

Comment (iemejia) [reply src/documentation/programming-guide.md]: @reuvenlax I am not talking about `ProcessContext` I refer to `@Element` and the new annotations not included in previous versions of 2.x. Users should be aware of this given that there is not backwards compatibility on those.

**Comment by reuvenlax** [comment]
@iemejia added a note that prior to Beam 2.5.0 ProcessContext should still be used.

**Comment by reuvenlax** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by kennknowles** [comment]
Noting that "Automatic staging of pull requests" is up to date while the "Jenkins: Automatic staging of pull requests" is older.

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
@reuvenlax Anything missing to merge this one ? Tests are ok given Kenn coment.

**Comment by iemejia** [comment]
Jumping on this one again. Anything else missing @reuvenlax / @melap ?

**Comment by reuvenlax** [comment]
Nothing missing to merge. I just was never able to get all tests to pass,
always for reasons unrelated to this change itself. I didn't see Kenn's
comment.

**Comment by melap** [comment]
I fixed a couple language tab tag problems and all looks good on staging now, merging

**Comment by melap** [comment]
@asfgit merge

**Pull request closed without merging**